File: apps/web-frontend/src/api/mock/report.ts
// Mock 大模型分析 + 报告 HTML 生成(后端未就绪时的演示)。
// 真后端:分析由 LLMClient 产出,报告由 WeasyPrint 渲染 PDF / python-docx 渲染 DOCX。
import type { AnalysisResult, AnalysisSection, DetectionJob } from '@/types'
import { formatDateTime, percent } from '@/utils/format'

const CLASS_CN: Record<string, string> = {
  glioma: '胶质瘤',
  meningioma: '脑膜瘤',
  pituitary: '垂体瘤',
}
const cn = (l: string) => CLASS_CN[l.toLowerCase()] || l

export function buildMockAnalysis(job: DetectionJob): AnalysisResult {
  const per = job.summary.per_class
  const entries = Object.entries(per).sort((a, b) => b[1] - a[1])
  const total = job.summary.count
  const withFinding = job.images.filter((i) => i.detections.length > 0).length
  const allConf = job.images.flatMap((i) => i.detections.map((d) => d.confidence))
  const avgConf = allConf.length
    ? allConf.reduce((a, b) => a + b, 0) / allConf.length
    : 0

  const sections: AnalysisSection[] = [
    {
      title: '总体概述',
      content:
        `本次共分析 MRI 影像 ${job.image_count} 例,其中 ${withFinding} 例检出疑似占位性病变,` +
        `累计检出目标 ${total} 处。所用模型为 ${job.model}。` +
        (entries.length
          ? `检出类别以${cn(entries[0][0])}为主(${entries[0][1]} 处)。`
          : '本批次未检出明显疑似病变区域。') +
        `整体检出置信度均值约 ${percent(avgConf)},结果仅供临床医师参考。`,
    },
    {
      title: '分类统计解读',
      content: entries.length
        ? entries
            .map(
              ([k, v]) =>
                `· ${cn(k)}(${k}):${v} 处,占检出总数的 ${percent(v / total, 0)}。`
            )
            .join('\n') +
          '\n上述分布反映本批次的检出构成,不代表确诊比例;同一患者多序列/多层面可能对同一病灶重复计数,需结合原始影像核对。'
        : '本批次未产生分类检出,建议结合临床表现与其他序列综合判断,必要时复查。',
    },
    {
      title: '置信度与不确定性说明',
      content:
        `检出置信度均值约 ${percent(avgConf)}。置信度反映模型对"该区域存在目标"的判断强度,` +
        '并非病变恶性程度或诊断确定性。对置信度偏低(如低于 0.5)的检出,假阳性风险较高,应重点复核;' +
        '对高置信度检出,也需人工确认边界与定位,避免漏诊邻近微小病灶。',
    },
    {
      title: '建议与后续',
      content:
        '1. 由具备资质的放射科/神经外科医师对上述检出逐一复核,结合 T1/T2/FLAIR/增强等多序列判断;\n' +
        '2. 对疑似病灶建议补充增强扫描或随访复查,明确性质与范围;\n' +
        '3. 结合患者临床症状、病史与实验室检查综合评估,制定进一步诊疗方案;\n' +
        '4. 本系统输出为辅助筛查结果,不作为独立诊断依据。',
    },
    {
      title: '免责声明',
      content:
        '本报告由人工智能辅助检测系统自动生成,所有检测与分析结果仅供临床参考,不能替代专业医师的诊断与临床判断。' +
        '任何诊疗决策应由具备资质的医师结合完整临床资料作出。系统开发方不对基于本报告作出的临床决策承担责任。',
    },
  ]

  return {
    provider: 'dashscope',
    llm_model: 'qwen-plus(mock)',
    sections,
    created_at: new Date().toISOString(),
  }
}

/** 生成企业风格的自包含 HTML 报告(可直接打印为 PDF) */
export function buildReportHtml(job: DetectionJob, analysis: AnalysisResult): string {
  const per = job.summary.per_class
  const entries = Object.entries(per).sort((a, b) => b[1] - a[1])
  const imgCards = job.images
    .map((im, i) => {
      const dets = im.detections
        .map(
          (d) =>
            `<tr><td>${cn(d.label)}</td><td>${d.label}</td><td class="mono">${(
              d.confidence * 100
            ).toFixed(1)}%</td><td class="mono">[${d.bbox.join(', ')}]</td></tr>`
        )
        .join('')
      return `
      <div class="img-block">
        <div class="img-title">影像 ${i + 1} · ${escapeHtml(im.filename)}
          <span class="muted">检出 ${im.detections.length} 处 · ${im.width}×${im.height}</span></div>
        <div class="img-pair">
          <figure><img src="${im.original_url}" /><figcaption>原始影像</figcaption></figure>
          <figure><img src="${im.result_url}" /><figcaption>检测标注</figcaption></figure>
        </div>
        ${
          im.detections.length
            ? `<table class="det"><thead><tr><th>类别</th><th>标签</th><th>置信度</th><th>坐标 [x1,y1,x2,y2]</th></tr></thead><tbody>${dets}</tbody></table>`
            : '<div class="no-det">未检出疑似病变区域</div>'
        }
      </div>`
    })
    .join('')

  const sectionsHtml = analysis.sections
    .map(
      (s) =>
        `<section class="analysis"><h3>${escapeHtml(s.title)}</h3><p>${escapeHtml(
          s.content
        ).replace(/\n/g, '<br/>')}</p></section>`
    )
    .join('')

  return `<!doctype html><html lang="zh-CN"><head><meta charset="utf-8"/>
<title>脑部 MRI 肿瘤检测报告 · ${job.id}</title>
<style>
  @page { size: A4; margin: 18mm 16mm; }
  * { box-sizing: border-box; }
  body { font-family: -apple-system,"Segoe UI","PingFang SC","Microsoft YaHei",sans-serif; color:#1f2733; font-size:12.5px; line-height:1.6; }
  .mono { font-family: "SFMono-Regular",Consolas,Menlo,monospace; }
  .muted { color:#909aa6; font-weight:400; font-size:11px; }
  header.rpt { display:flex; justify-content:space-between; align-items:flex-start; border-bottom:2px solid #0d8b8a; padding-bottom:12px; margin-bottom:16px; }
  .brand { display:flex; align-items:center; gap:10px; }
  .brand .logo { width:34px; height:34px; border-radius:8px; background:#0d8b8a; color:#fff; display:flex; align-items:center; justify-content:center; font-weight:700; }
  .brand h1 { font-size:16px; margin:0; }
  .brand p { margin:2px 0 0; color:#56606d; font-size:11px; }
  .meta { text-align:right; font-size:11px; color:#56606d; }
  .summary { display:flex; gap:10px; margin:14px 0 18px; }
  .kpi { flex:1; border:1px solid #e6e9ef; border-radius:8px; padding:10px 12px; background:#fafbfc; }
  .kpi .n { font-size:20px; font-weight:700; color:#0d8b8a; }
  .kpi .l { font-size:11px; color:#56606d; }
  h2.sec { font-size:14px; margin:20px 0 10px; padding-left:9px; border-left:3px solid #0d8b8a; }
  table { width:100%; border-collapse:collapse; margin-top:6px; }
  th,td { border:1px solid #e6e9ef; padding:5px 8px; text-align:left; font-size:11.5px; }
  th { background:#f4f6f9; }
  .img-block { margin-bottom:16px; page-break-inside:avoid; }
  .img-title { font-weight:600; font-size:12.5px; margin-bottom:6px; }
  .img-pair { display:flex; gap:10px; }
  .img-pair figure { flex:1; margin:0; text-align:center; }
  .img-pair img { width:100%; border:1px solid #24304a; border-radius:6px; background:#0b1220; }
  .img-pair figcaption { font-size:10.5px; color:#56606d; margin-top:3px; }
  .no-det { color:#909aa6; font-size:11.5px; padding:6px 0; }
  section.analysis { margin-bottom:10px; page-break-inside:avoid; }
  section.analysis h3 { font-size:12.5px; margin:0 0 4px; color:#0a6e6d; }
  section.analysis p { margin:0; }
  .disclaimer { margin-top:16px; padding:10px 12px; background:#fff6ee; border:1px solid #f2d5b8; border-radius:8px; font-size:11px; color:#8a5a2b; }
  footer.rpt { margin-top:18px; border-top:1px solid #e6e9ef; padding-top:8px; font-size:10px; color:#909aa6; text-align:center; }
</style></head><body>
<header class="rpt">
  <div class="brand">
    <div class="logo">OD</div>
    <div><h1>脑部 MRI 肿瘤检测报告</h1><p>AI 辅助影像检测 · 仅供临床参考</p></div>
  </div>
  <div class="meta">
    报告编号:${job.id}<br/>
    生成时间:${formatDateTime(analysis.created_at)}<br/>
    检测模型:${escapeHtml(job.model)}<br/>
    分析引擎:${escapeHtml(analysis.llm_model)}
  </div>
</header>

<div class="summary">
  <div class="kpi"><div class="n">${job.image_count}</div><div class="l">影像数量</div></div>
  <div class="kpi"><div class="n">${job.summary.count}</div><div class="l">检出目标总数</div></div>
  <div class="kpi"><div class="n">${entries.length}</div><div class="l">涉及类别</div></div>
  <div class="kpi"><div class="n">${entries[0] ? cn(entries[0][0]) : '—'}</div><div class="l">主要类别</div></div>
</div>

<h2 class="sec">一、智能分析</h2>
${sectionsHtml}

<h2 class="sec">二、逐例影像检测</h2>
${imgCards}

<div class="disclaimer">⚠️ 免责声明:本报告由 AI 辅助检测系统自动生成,检测与分析结果仅供临床参考,<b>不能替代专业医师的诊断</b>。任何诊疗决策应由具备资质的医师结合完整临床资料作出。</div>

<footer class="rpt">ODPlatform 脑部 MRI 肿瘤检测系统 · 本页由系统自动生成</footer>
</body></html>`
}

function escapeHtml(s: string): string {
  return s
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
}

File: apps/web-backend/src/od_web_backend/analysis/report.py
"""报告渲染:企业风格 HTML → PDF(WeasyPrint 优先,缺系统库回退 xhtml2pdf)+ DOCX(python-docx)。

图片以 base64 内联进 HTML,PDF 两种引擎都能出图;DOCX 直接按文件路径插图。
"""
from __future__ import annotations
import base64
import logging
from io import BytesIO
from pathlib import Path

from django.conf import settings

from .prompts import cn

logger = logging.getLogger(__name__)


def _abs(rel: str) -> Path:
    return Path(settings.MEDIA_ROOT) / rel


def _data_uri(rel: str) -> str:
    try:
        p = _abs(rel)
        b = p.read_bytes()
        return 'data:image/jpeg;base64,' + base64.b64encode(b).decode('ascii')
    except Exception:
        return ''


def _esc(s: str) -> str:
    return (str(s).replace('&', '&amp;').replace('<', '&lt;')
            .replace('>', '&gt;').replace('"', '&quot;'))


def _fmt_dt(iso: str) -> str:
    return (iso or '').replace('T', ' ')[:16]


# --------------------------------------------------------------------------- #
# HTML(表格布局,兼容 xhtml2pdf 与 WeasyPrint)
# --------------------------------------------------------------------------- #
def build_report_html(job, analysis: dict) -> str:
    summary = job.summary or {}
    per = summary.get('per_class', {})
    entries = sorted(per.items(), key=lambda kv: kv[1], reverse=True)
    images = list(job.images.all().prefetch_related('detections'))

    img_blocks = []
    for i, im in enumerate(images):
        dets = list(im.detections.all())
        rows = ''.join(
            f'<tr><td>{cn(d.label)}</td><td>{d.label}</td>'
            f'<td class="mono">{d.confidence * 100:.1f}%</td>'
            f'<td class="mono">[{", ".join(map(str, d.bbox))}]</td></tr>'
            for d in dets
        )
        det_table = (
            f'<table class="det"><thead><tr><th>类别</th><th>标签</th><th>置信度</th>'
            f'<th>坐标 [x1,y1,x2,y2]</th></tr></thead><tbody>{rows}</tbody></table>'
            if dets else '<div class="no-det">未检出疑似病变区域</div>'
        )
        img_blocks.append(f'''
        <div class="img-block">
          <div class="img-title">影像 {i + 1} · {_esc(im.filename)}
            <span class="muted">检出 {len(dets)} 处 · {im.width}×{im.height}</span></div>
          <table class="pair"><tr>
            <td><img src="{_data_uri(im.original)}"/><div class="cap">原始影像</div></td>
            <td><img src="{_data_uri(im.result)}"/><div class="cap">检测标注</div></td>
          </tr></table>
          {det_table}
        </div>''')

    sections_html = ''.join(
        f'<div class="analysis"><h3>{_esc(s["title"])}</h3>'
        f'<p>{_esc(s["content"]).replace(chr(10), "<br/>")}</p></div>'
        for s in analysis.get('sections', [])
    )

    return f'''<!doctype html><html lang="zh-CN"><head><meta charset="utf-8"/>
<style>
  @page {{ size: A4; margin: 1.6cm 1.4cm; }}
  body {{ font-family: cjk, "Microsoft YaHei", "PingFang SC", sans-serif; color:#1f2733; font-size:12px; }}
  .mono {{ font-family: cjk, monospace; }}
  .muted {{ color:#909aa6; font-weight:normal; font-size:10px; }}
  .header {{ border-bottom:2px solid #0d8b8a; padding-bottom:8px; margin-bottom:12px; }}
  .header h1 {{ font-size:17px; margin:0 0 4px; color:#0d8b8a; }}
  .header .sub {{ color:#56606d; font-size:11px; }}
  .meta {{ font-size:10.5px; color:#56606d; margin-top:6px; }}
  .kpis {{ width:100%; margin:10px 0 14px; }}
  .kpis td {{ border:1px solid #e6e9ef; background:#fafbfc; padding:8px 10px; text-align:center; width:25%; }}
  .kpis .n {{ font-size:18px; font-weight:bold; color:#0d8b8a; }}
  .kpis .l {{ font-size:10px; color:#56606d; }}
  h2.sec {{ font-size:13px; color:#0a6e6d; border-left:3px solid #0d8b8a; padding-left:8px; margin:16px 0 8px; }}
  .analysis {{ margin-bottom:8px; }}
  .analysis h3 {{ font-size:12px; margin:0 0 3px; color:#0a6e6d; }}
  .analysis p {{ margin:0; line-height:1.6; color:#333; }}
  .img-block {{ margin-bottom:14px; }}
  .img-title {{ font-weight:bold; font-size:12px; margin-bottom:5px; }}
  table.pair {{ width:100%; }}
  table.pair td {{ width:50%; text-align:center; vertical-align:top; padding:2px; }}
  table.pair img {{ width:100%; border:1px solid #24304a; }}
  .cap {{ font-size:10px; color:#56606d; margin-top:2px; }}
  table.det {{ width:100%; border-collapse:collapse; margin-top:5px; }}
  table.det th, table.det td {{ border:1px solid #e6e9ef; padding:4px 6px; font-size:10.5px; text-align:left; }}
  table.det th {{ background:#f4f6f9; }}
  .no-det {{ color:#909aa6; font-size:11px; padding:4px 0; }}
  .disclaimer {{ margin-top:14px; padding:8px 10px; background:#fff6ee; border:1px solid #f2d5b8; font-size:10.5px; color:#8a5a2b; }}
  .footer {{ margin-top:14px; border-top:1px solid #e6e9ef; padding-top:6px; font-size:9.5px; color:#909aa6; text-align:center; }}
</style></head><body>
<div class="header">
  <h1>脑部 MRI 肿瘤检测报告</h1>
  <div class="sub">AI 辅助影像检测 · 仅供临床参考</div>
  <div class="meta">报告编号:{job.id} &nbsp;|&nbsp; 生成时间:{_fmt_dt(analysis.get("created_at", ""))}
    &nbsp;|&nbsp; 检测模型:{_esc(job.model)} &nbsp;|&nbsp; 分析引擎:{_esc(analysis.get("llm_model", ""))}</div>
</div>

<table class="kpis"><tr>
  <td><div class="n">{job.image_count}</div><div class="l">影像数量</div></td>
  <td><div class="n">{summary.get("count", 0)}</div><div class="l">检出目标总数</div></td>
  <td><div class="n">{len(entries)}</div><div class="l">涉及类别</div></td>
  <td><div class="n">{cn(entries[0][0]) if entries else "—"}</div><div class="l">主要类别</div></td>
</tr></table>

<h2 class="sec">一、智能分析</h2>
{sections_html}

<h2 class="sec">二、逐例影像检测</h2>
{''.join(img_blocks)}

<div class="disclaimer">⚠️ 免责声明:本报告由 AI 辅助检测系统自动生成,检测与分析结果仅供临床参考,<b>不能替代专业医师的诊断</b>。任何诊疗决策应由具备资质的医师结合完整临床资料作出。</div>
<div class="footer">ODPlatform 脑部 MRI 肿瘤检测系统 · 本页由系统自动生成</div>
</body></html>'''


# --------------------------------------------------------------------------- #
# PDF
# --------------------------------------------------------------------------- #
def _find_cjk_font() -> str | None:
    if settings.VIZ_FONT_PATH and Path(settings.VIZ_FONT_PATH).exists():
        return settings.VIZ_FONT_PATH
    for c in [
        '/usr/share/fonts/opentype/noto/NotoSansCJK-Regular.ttc',
        '/usr/share/fonts/opentype/noto/NotoSansCJK-Black.ttc',
        '/usr/share/fonts/truetype/wqy/wqy-zenhei.ttc',
        '/usr/share/fonts/truetype/arphic/ukai.ttc',
    ]:
        if Path(c).exists():
            return c
    return None


def _register_cjk_xhtml2pdf() -> None:
    """给 xhtml2pdf(reportlab)注册 CJK 字体,让中文能渲染。"""
    font_path = _find_cjk_font()
    if not font_path:
        return
    try:
        from reportlab.pdfbase import pdfmetrics
        from reportlab.pdfbase.ttfonts import TTFont
        if 'cjk' not in pdfmetrics.getRegisteredFontNames():
            try:
                pdfmetrics.registerFont(TTFont('cjk', font_path))
            except Exception:
                pdfmetrics.registerFont(TTFont('cjk', font_path, subfontIndex=0))
    except Exception as exc:
        logger.warning('CJK 字体注册失败(xhtml2pdf): %s', exc)


def render_pdf(html: str) -> bytes:
    engine = settings.REPORT_PDF_ENGINE
    if engine in ('auto', 'weasyprint'):
        try:
            from weasyprint import HTML  # 需系统库(cairo/pango)
            return HTML(string=html).write_pdf()
        except Exception as exc:
            if engine == 'weasyprint':
                raise
            logger.info('WeasyPrint 不可用,回退 xhtml2pdf: %s', exc)

    from xhtml2pdf import pisa
    _register_cjk_xhtml2pdf()
    out = BytesIO()
    pisa.CreatePDF(src=html, dest=out, encoding='utf-8')
    return out.getvalue()


# --------------------------------------------------------------------------- #
# DOCX
# --------------------------------------------------------------------------- #
def render_docx(job, analysis: dict) -> bytes:
    from docx import Document
    from docx.shared import Pt, RGBColor, Inches
    from docx.enum.text import WD_ALIGN_PARAGRAPH

    summary = job.summary or {}
    per = summary.get('per_class', {})
    entries = sorted(per.items(), key=lambda kv: kv[1], reverse=True)
    images = list(job.images.all().prefetch_related('detections'))

    doc = Document()
    style = doc.styles['Normal']
    style.font.name = 'Microsoft YaHei'
    style.font.size = Pt(10.5)

    title = doc.add_heading('脑部 MRI 肿瘤检测报告', level=0)
    for run in title.runs:
        run.font.color.rgb = RGBColor(0x0D, 0x8B, 0x8A)
    doc.add_paragraph('AI 辅助影像检测 · 仅供临床参考')

    meta = doc.add_paragraph()
    meta.add_run(
        f'报告编号:{job.id}\n生成时间:{_fmt_dt(analysis.get("created_at", ""))}\n'
        f'检测模型:{job.model}\n分析引擎:{analysis.get("llm_model", "")}'
    ).font.size = Pt(9)

    # KPI 表
    kt = doc.add_table(rows=2, cols=4)
    kt.style = 'Light Grid Accent 1'
    headers = ['影像数量', '检出目标总数', '涉及类别', '主要类别']
    values = [str(job.image_count), str(summary.get('count', 0)), str(len(entries)),
              cn(entries[0][0]) if entries else '—']
    for j, (h, v) in enumerate(zip(headers, values)):
        kt.cell(0, j).text = h
        kt.cell(1, j).text = v

    # 分析
    doc.add_heading('一、智能分析', level=1)
    for s in analysis.get('sections', []):
        doc.add_heading(s['title'], level=2)
        doc.add_paragraph(s['content'])

    # 逐例影像
    doc.add_heading('二、逐例影像检测', level=1)
    for i, im in enumerate(images):
        dets = list(im.detections.all())
        doc.add_heading(f'影像 {i + 1} · {im.filename}(检出 {len(dets)} 处)', level=2)
        # 原图 + 标注图
        pt = doc.add_table(rows=1, cols=2)
        for cell, rel, cap in [(pt.cell(0, 0), im.original, '原始影像'),
                               (pt.cell(0, 1), im.result, '检测标注')]:
            para = cell.paragraphs[0]
            para.alignment = WD_ALIGN_PARAGRAPH.CENTER
            try:
                run = para.add_run()
                run.add_picture(str(_abs(rel)), width=Inches(2.7))
            except Exception:
                para.add_run('(图片缺失)')
            cap_p = cell.add_paragraph(cap)
            cap_p.alignment = WD_ALIGN_PARAGRAPH.CENTER
            cap_p.runs[0].font.size = Pt(8)
        if dets:
            dt = doc.add_table(rows=1, cols=4)
            dt.style = 'Light List Accent 1'
            for j, h in enumerate(['类别', '标签', '置信度', '坐标']):
                dt.cell(0, j).text = h
            for d in dets:
                row = dt.add_row().cells
                row[0].text = cn(d.label)
                row[1].text = d.label
                row[2].text = f'{d.confidence * 100:.1f}%'
                row[3].text = f'[{", ".join(map(str, d.bbox))}]'
        else:
            doc.add_paragraph('未检出疑似病变区域')

    doc.add_paragraph()
    disc = doc.add_paragraph()
    r = disc.add_run(
        '⚠️ 免责声明:本报告由 AI 辅助检测系统自动生成,检测与分析结果仅供临床参考,'
        '不能替代专业医师的诊断。任何诊疗决策应由具备资质的医师结合完整临床资料作出。'
    )
    r.font.size = Pt(9)
    r.font.color.rgb = RGBColor(0x8A, 0x5A, 0x2B)

    buf = BytesIO()
    doc.save(buf)
    return buf.getvalue()
